README.md
# REMORA

![REMORA Logo](https://raw.githubusercontent.com/TACC/remora/master/docs/logos/Remora-logo-300px.png)

## REMORA: REsource MOnitoring for Remote Applications

Remora is a tool to monitor runtime resource utilization:
  - Memory
  - CPU utilization
  - IO usage (Lustre, DVS)
  - NUMA memory
  - Network topology
  - MPI statistics
  - CPU Power and Temperature data

To use the tool, modify your batch script and include 'remora' before your script, executable, or MPI launcher.



## Download and use
Please, **do not try to use the version available in the master branch**. We regularly change the code and it might contain bugs. If you want to download and use remora, **have a look at the different tags**. [The most recent release can be found here](https://github.com/TACC/remora/releases/latest).

## Why name it Remora?
Apart from a pretty cool acronym, this tools behaves a bit like the remora fish. It attaches to a larger fish (user process) and travels with it wherever it goes, while offering very little in the way of resistance to the motion (overhead) as well as providing some benefits (resource usage information).

## Do you use Remora?
Remora is an open-source project. Funding to keep researchers working on Remora depends on the value of this tool to the scientific community. We would appreciate if you could include the following citation in your scientific articles:

[C. Rosales, A. Gómez-Iglesias, A. Predoehl. "REMORA: a Resource Monitoring Tool for Everyone". HUST2015 November 15-20, 2015, Austin, TX, USA. DOI: 10.1145/2834996.2834999](http://dl.acm.org/citation.cfm?id=2834999)

## Comments? Suggestions?
Feel free to create new issues here in GitHub. You can also send us an email.


ChangeLog.md
# Change Log

## [1_8_4] - (2020-l0-19)
* MPI
  1.) Reporting MPI functions was fixed at 5. (Broke codes with less.) (Fixed)
      Changed to gathering top 20 or all greater than 0.5%.
  2.) Long jobs broke because mpiP stats returned time in sci. notation. (Fixed)
  3.) MPI fraction for mpiP now reports coefficient of variation.
  4.) Changed reporting values of MAX (%) to Average.
  5.) Fixed recent change to broke reporting memory in summary for multi-nodes runs

*NUMA
  1.) Incorrectly stated 4MB pages instead of 2MB. (Fixed)

*GPU
  1.) On dual system (nodes with/without GPU) report GPU is active, but not
      used on non-GPU nodes.
  2.) No longer necessary to set REMORA_CUDA to get GPU information on gpu nodes.
      Can turn off by setting REMORA_CUD to any value except 1.

*MEM
  1.) Summary now reports MAX VIRT, MAX PHYS and AVAIL at beginning of job.
  

## [1_8_3] - (2019-04-25)

### Codes Fixes
* numa
  1. Corrected 4MB to 2MB in description of pages.

* remora_finalize.sh
  1. Create local nodes array, NODES_LIST, for accessing node and pid simultaneous as list
     (e.g.  ${PID[$idx]} and ${NODES_LIST[$idx]}  in loop)
* install.sh
   1. config/modules must be changed from ib,NETWORK opa,NETWORK if OPA is detected

* scripts/remora_finalize.sh
   1. Include impi fraction and breakdown links under "impi utilization".
   2. impi_fraction.html and impi_breakdown.html links in remora_summary.html were not
   3. included. Now, it is included if impi is found in REMORA_MODULES[@] in remora_finalize.sh.
   4. (This works similar to including the lustre_aggregated.html under "lustre utilization".

* modules/numa
   1. numastats does not show THPs. 
   2. Replace misses with foreign hits--since it is the real miss. 
   3. Now uses vmstats to give Top-Right plot with allocated memory from thp and (thp+4k) pages. How to read text was placed at top.

* modules/opa
   1. Fixed modules/opa.  
   2. packs and bytes strings were never initialized.  pack was used in string names instead of packs.

* modules/network
   1. ibtracert not found in default path.  
   2. Added /sbin

* modules/ib
   1. was not set up to handle dev_hfi1_0 device for opa

* docs
   1. added modules_help and modules_whatis files for supporting mkmod



## [1_8_2] - (2017-08-08)

Small issue fix (#45). Also ensures that there are extra checks when no scheduler is present.

## [1_8_1] - (2017-06-30)

Several issues fixed (#37, #40, #41, #42, #43) and partial fix for issue #39.
For #39, the CPU collection is now correct when mpstats is not in the path of all nodes

## [1_8_0] - (2017-03-09)

Added MPI statistics collection for Intel MPI and Mvapich2
Modified how top level functions are called in order to improve overall handling
Improved verbose mode information

## [1_7_1] - (2016-12-07)

Now collection and post-processing work for users with default csh/tsh shells.

Thanks to Mark Reed (UNC) for the suggested changes.

## [1_7_0] - (2016-12-01)

Added power and temperature modules.
Added automatic discovery of NUMA nodes.
Supports PBS scheduler.
Supports running on a local machine (no scheduler needed).
Multiple jobs for the same run: new folders are created.
Use Google Charts to create plots automatically (no need for post-processing Python script except for network trace).
Added a script to monitor free memory that kill the application before the OOM does.
Added a script to generate post-processing files in case the node crashes.

Contributions from Jacob Pollack at Shell (Eth monitoring, CPU utilization post-processing, and 64-bit counters for Infiniband monitoring), Kevin Manalo (suggestion of automation of mlx device discovery) and wpoely86 (PBS support).

## [bp-1_6_0] - (2016-08-09)

Added "eth" module which tracks Ethernet statistics in a similar fashion to how "ib" module tracks InfiniBand statistics.
Added support for use of 64 bit InfiniBand counters if they exist.
Added average CPU usage information to CPU plots.

## [1_6_0] - (2016-03-11)

Introduced real-time monitoring capabilities, enabled by REMORA_MODE=MONITOR.
Fixed minor bug introduced in summary report warnings for systems with over 32 GB or memory per node.
Updated user documentation.

## [1_5_0] - (2016-02-16)

Improved performance by allowing the use of a separate REMORA_TMPDIR which can be a local file system.
Simplified "remora" script itself by having a remora_init / remora_collect / remora_finalize scripts.
New "src/scripts" directory to clean source structure.
Cleaned up output directory and improved summary formatting.
Using "remora_env.txt" file in shared location REMORA_OUTDIR to capture remora options during runtime.  
Added DVS file system data collection and file system blacklist.

## [1_4_0] - 2016-01-25

Improved the format of the output files. The columns have a fixed width so that they are easier to read.
Implemented a modular design, where all the data captured by REMORA is done by independent scripts in the
modules folder. A configuration file, that needs to be configured as part of the installation process,
specifies the active modules for a particular site.

## [1_3_1] - 2015-10-21

Changed remora_post to work with older versions of Python. It was only working with Python 2.7+

## [1_3_0] - 2015-09-19

Added memory monitoring for NVIDIA GPUs

## [1_2_0] - 2015-09-11

Minor bugfixes

## [1_1_0] - 2015-09-09

Fixed Lustre data parser

## [1_0_0] - 2015-08-18

First public release of REMORA (REsource MOnitoring for Remote Applications)
Remora is a tool to monitor runtime resource utilization:

* Memory
* CPU utilization
* Lustre usage
* NUMA memory
* Network topology

To use the tool, modify your batch script and include 'remora' before your script, executable, or MPI launcher.


install.sh
#!/bin/sh
# Installation script for REMORA
#
# Change CC, MPICC and the corresponding flags to match your own compiler in
# file "Makefile.in". You should not have to edit this file at all.
#
# v1_8_4 (2020-10-19)  Kent Milfeld
# v1_8_3 (2018-04-25)  Kent Milfeld/Si Liu
# v1_8_2 (2017-08-08)  Carlos Rosales-Fernandez
#                      Antonio Gomez-Iglesias
#
# Thanks to Kenneth Hoste, from HPC-UGent, for his input
# Changed to use mpiP for intel 2019 and above  (impi_mpiP) (Kent)

# installation directory: use $REMORA_INSTALL_PREFIX if defined, current directory if not
export REMORA_DIR=${REMORA_INSTALL_PREFIX:-$PWD}
export PHI_BUILD=0

# Do not change anything below this line
#--------------------------------------------------------------------------------

mkdir -p $REMORA_DIR/bin
mkdir -p $REMORA_DIR/include
mkdir -p $REMORA_DIR/lib
mkdir -p $REMORA_DIR/share
mkdir -p $REMORA_DIR/docs

REMORA_BUILD_DIR=$PWD

VERSION=1.8.4
COPYRIGHT1="Copyright 2017 The University of Texas at Austin."
COPYRIGHT2="License: MIT <http://opensource.org/licenses/MIT>"
COPYRIGHT3="This is free software: you are free to change and redistribute it."
COPYRIGHT4="There is NO WARRANTY of any kind"

BUILD_LOG="$REMORA_BUILD_DIR/remora_build.log"
INSTALL_LOG="$REMORA_BUILD_DIR/remora_install.log"

SEPARATOR="======================================================================"
   PKG="Package  : REMORA"
   VER="Version  : $VERSION"
  DATE="Date     : `date +%Y.%m.%d`"
SYSTEM="System   : `uname -sr`"

# Record the local conditions for the compilation
echo
echo $SEPARATOR  | tee $BUILD_LOG
echo $PKG        | tee -a $BUILD_LOG
echo $VER        | tee -a $BUILD_LOG
echo $DATE       | tee -a $BUILD_LOG
echo $SYSTEM     | tee -a $BUILD_LOG
echo $SEPARATOR  | tee -a $BUILD_LOG
echo $COPYRIGHT1 | tee -a $BUILD_LOG
echo $COPYRIGHT2 | tee -a $BUILD_LOG
echo $COPYRIGHT3 | tee -a $BUILD_LOG
echo $COPYRIGHT4 | tee -a $BUILD_LOG
echo $SEPARATOR  | tee -a $BUILD_LOG
echo             | tee -a $BUILD_LOG

#Now build mpstat
echo "Building mpstat ..." | tee -a $BUILD_LOG
cd $REMORA_BUILD_DIR/extra
sysfile=`ls -ld sysstat*.tar.gz | awk '{print $9}' | head -n 1`
sysdir=`echo  ${sysfile%%.tar.gz}`
tar xzvf ${sysfile}
cd ${sysdir}
./configure | tee -a $BUILD_LOG
make mpstat |  tee -a $BUILD_LOG
echo "Installing mpstat ..." | tee -a $INSTALL_LOG
cp mpstat $REMORA_DIR/bin

#Now build mpiP
# Check if MPI compiler is available. If not, disable MPI module and skip mpiP build
export CC=mpicc
export F77=mpif77
export ARCH=x86_64

if [ "$( $CC --version >& /dev/null || echo "0")" == "0" ]; then 
    haveMPICC=0
else 
    haveMPICC=1
fi

if [ "$( $F77 --version >& /dev/null || echo "0")" == "0" ]; then
     haveMPIFC=0
else 
     haveMPIFC=1 
fi

if [[ $haveMPICC == 1 && $haveMPIFC == 1 ]]; then

   mpicc -v |& grep 'Intel(R) MPI' >& /dev/null
   haveIMPI=$((1-${PIPESTATUS[1]}))           #1=has IMPI, 0=does not have IMPI

   mpicc -v |& grep 'MVAPICH2'     >& /dev/null
   haveMV2=$((1-${PIPESTATUS[1]}))            #1=has MVAPICH2, 0=does not have MV2

   if [[ $haveIMPI == 1 ]]; then
      IMPI_year=$( mpicc -v |& grep 'Library 20' | sed 's/.*Library 20\([0-9][0-9]\).*/\1/' )
      if [[ $IMPI_year > 18 ]]; then
         IMPI_stats="impi_mpip"
      else
         IMPI_stats=impi
      fi
      #mpicc -v |& grep 'MPI Library 2019' >& /dev/null
      #haveIMPI19=$((1-${PIPESTATUS[1]}))      #1=has IMPI 19, 0=does not have IMPI 19
   fi
else
    echo ""
    [[ $haveMPICC == 0 ]] && echo " WARNING : mpicc not found " | tee -a $BUILD_LOG
    [[ $haveMPIFC == 0 ]] && echo " WARNING : mpi77 not found " | tee -a $BUILD_LOG
    echo " WARNING : REMORA will be built without MPI support " | tee -a $BUILD_LOG
    echo ""
fi


if [[ "$haveMPICC" == "1" && "$haveMPIFC"  == "1" ]]; then

   if [[ "$haveMV2"    == "1" || "$IMPI_stats" == "impi_mpip" ]] ; then

      [[ "$haveMV2"    == "1" ]]         && echo " REMORA built with mpiP statistics for Mvapich2"  | tee -a $BUILD_LOG
      [[ "$IMPI_stats" == "impi_mpip" ]] && echo " REMORA built with mpiP statistics for IMPI 20$IMPI_year" | tee -a $BUILD_LOG

      echo "Building mpiP ..." | tee -a $BUILD_LOG
      cd $REMORA_BUILD_DIR/extra
      mpipfile=`ls -ld mpiP*.tar.gz | awk '{print $9}' | head -n 1`
      mpipdir=`echo  ${mpipfile%%.tar.gz}`
      tar xzvf ${mpipfile}
      cd ${mpipdir}
      ./configure CFLAGS="-g" --enable-demangling --disable-bfd --disable-libunwind --prefix=${REMORA_DIR} | tee -a $BUILD_LOG
      make        | tee -a $BUILD_LOG
      make shared | tee -a $BUILD_LOG
      echo "Installing mpiP ..." | tee -a $INSTALL_LOG
      make install

   fi

   if [[ "$IMPI_stats" == "impi" ]]; then
      echo "" 
      echo " REMORA built with support for impi (ipm) statistics" | tee -a $BUILD_LOG
      echo ""
   fi

   if [[ "$haveMV2" == "0" && "$haveIMPI" == "0" ]]; then
      echo "" 
      echo " REMORA only supports statistics for IMPI and MVAPICH2"          | tee -a $BUILD_LOG
      echo " WARNING : REMORA will be built with NO MPI statistics support " | tee -a $BUILD_LOG
      echo ""
   fi

fi

if [ "$PHI_BUILD" == "1" ]; then
	echo "Building Xeon Phi affinity script ..."   |  tee -a $BUILD_LOG
	cd $REMORA_BUILD_DIR/extra/
	icc -mmic -o ma ./mic_affinity.c
	echo "Installing Xeon Phi affinity script ..." |  tee -a $INSTALL_LOG
	cp -v ./ma $REMORA_DIR/bin                     |  tee -a $INSTALL_LOG
fi

echo "Copying all scripts to installation folder ..." |  tee -a $INSTALL_LOG

cd $REMORA_BUILD_DIR
cp -vr ./src/* $REMORA_DIR/bin

if [ "$haveMPICC" == "0" ] || [ "$haveMPIFC" == "0" ]; then
    sed -i '/impi,MPI/d'      $REMORA_DIR/bin/config/modules
    sed -i '/impi_mpip,MPI/d' $REMORA_DIR/bin/config/modules
    sed -i '/mv2,MPI/d'       $REMORA_DIR/bin/config/modules
else
    if [ "$haveIMPI" == "1" ]; then
                                           sed -i '/mv2,MPI/d'       $REMORA_DIR/bin/config/modules
       [[ $IMPI_stats == "impi_mpip" ]] && sed -i '/impi,MPI/d'      $REMORA_DIR/bin/config/modules
       [[ $IMPI_stats == "impi"      ]] && sed -i '/impi_mpip,MPI/d' $REMORA_DIR/bin/config/modules
    fi
    if [ "$haveMV2" == "1" ]; then
       sed -i '/impi,MPI/d'      $REMORA_DIR/bin/config/modules
       sed -i '/impi_mpip,MPI/d' $REMORA_DIR/bin/config/modules
    fi
fi

if [ "$haveMPICC" == "1" ] && [ "$haveMPIFC" == "1" ]; then
   /sbin/lsmod | grep hfi1 >& /dev/null
   if [[ $? == 0 ]]; then
      sed -i 's/ib,NETWORK/opa,NETWORK/' $REMORA_DIR/bin/config/modules
   fi
fi

if [[ "$REMORA_BUILD_DIR/docs" != $REMORA_DIR/docs ]]; then
        cp $PWD/docs/modules_help          $REMORA_DIR/docs
        cp $PWD/docs/modules_whatis        $REMORA_DIR/docs
        cp $PWD/docs/remora_user_guide.pdf $REMORA_DIR/docs
  echo "Copied modules_help modules_whatis and remora.pdf to $REMORA_DIR/docs dir."
fi

echo $SEPARATOR
echo "Installation of REMORA v$VERSION completed."
echo "For a fully functional installation make sure to:"; echo ""
echo "	export PATH=\$PATH:$REMORA_DIR/bin"
echo "	export LD_LIBRARY_PATH=\$LD_LIBRARY_PATH:$REMORA_DIR/lib"
echo "	export REMORA_BIN=$REMORA_DIR/bin"; echo ""
echo "Good Luck!"
echo $SEPARATOR
